tests/integration/test_db_router.py

```diff
@@ -171,60 +109,79 @@ async def check_keystone_users_existence(
-        status="active",
-        raise_on_blocked=True,
-        timeout=FAST_WAIT_TIMEOUT,
-    )
-    assert (
-        len(ops_test.model.applications[APP_NAME].units) == units
-    ), "Not all mysql units deployed"
 
-    for unit in ops_test.model.applications[APP_NAME].units:
-        assert unit.workload_status == "active", "Unit is not active"
+    mysql_app = await ops_test.model.deploy(
+        charm, application_name=APP_NAME, config=config, num_units=1
+    )
 
-    # Get the server config credentials
-    db_unit = ops_test.model.applications[APP_NAME].units[0]
-    return await get_server_config_credentials(db_unit)
+    # Deploy keystone
+    # Explicitly setting the series to 'focal' as it defaults to 'xenial'
+    keystone_app = await ops_test.model.deploy(
+        "keystone",
+        series="focal",
+        application_name=KEYSTONE_APP_NAME,
+        num_units=2,
+    )
 
+    # Deploy mysqlrouter and relate it to keystone
+    keystone_mysqlrouter_app = await ops_test.model.deploy(
+        "mysql-router",
+        channel="8.0/stable",  # pin to channel as it contains a fix to https://bugs.launchpad.net/charm-mysql-router/+bug/1927981
+        application_name=KEYSTONE_MYSQLROUTER_APP_NAME,
+    )
 
-@pytest.mark.order(1)
-@pytest.mark.abort_on_fail
-@pytest.mark.db_router_tests
-async def test_keystone_bundle_db_router(ops_test: OpsTest) -> None:
-    """Deploy the keystone bundle to test the 'db-router' relation.
+    await ops_test.model.relate(
+        f"{KEYSTONE_APP_NAME}:shared-db",
+        f"{KEYSTONE_MYSQLROUTER_APP_NAME}:shared-db",
+    )
 
-    Args:
-        ops_test: The ops test framework
-    """
     # Reduce the update_status frequency for the duration of the test
     async with ops_test.fast_forward():
 
-        result = await asyncio.gather(
-            build_and_deploy_mysql(ops_test, 3),
-            # Deploy and test the first deployment of keystone
-            deploy_and_relate_keystone_with_mysqlrouter(
-                ops_test, KEYSTONE_APP_NAME, KEYSTONE_MYSQLROUTER_APP_NAME, 2
+        await asyncio.gather(
+            ops_test.model.block_until(
+                lambda: mysql_app.status in ("active", "error"), timeout=SLOW_WAIT_TIMEOUT
             ),
+            ops_test.model.block_until(
+                lambda: keystone_app.status in ("waiting", "error"), timeout=SLOW_WAIT_TIMEOUT
+            ),
+            ops_test.model.block_until(
+                lambda: keystone_mysqlrouter_app.status in ("blocked", "error"),
+                timeout=SLOW_WAIT_TIMEOUT,
+            ),
+        )
+        assert (
+            mysql_app.status == "active"
+            and keystone_app.status == "waiting"
+            and keystone_mysqlrouter_app.status == "blocked"
+        )
+
+        # Relate mysqlrouter to mysql
+        await ops_test.model.relate(
+            f"{KEYSTONE_MYSQLROUTER_APP_NAME}:db-router", f"{APP_NAME}:db-router"
+        )
+        await ops_test.model.block_until(
```

Comment:
you could use wait_for_idle(), with `status="error"` and `block_on_error=False`. Up to you though

Reply:
we did use `wait_for_idle` before, but it apparently takes longer to stabilize after scaling up. so we switched to `block_until` because the only thing we care about is whether the application went to `active` status